File: 07_Trees/04_Balanced_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 * int val;
 * TreeNode *left;
 * TreeNode *right;
 * TreeNode() : val(0), left(nullptr), right(nullptr) {}
 * TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 * TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */

class Solution
{
public:
    int dfs(TreeNode *root)
    {
        if (!root)
            return 0;
        int left = dfs(root->left);
        int right = dfs(root->right);
        if (left == -1 || right == -1 || abs(left - right) > 1)
            return -1;
        return max(left, right) + 1;
    }

    bool isBalanced(TreeNode *root)
    {
        return dfs(root) != -1;
    }
};

class Solution
{
public:
    pair<int, bool> checkSubtreeDepth(TreeNode *root)
    {
        if (!root)
            return {0, true};
        auto left = checkSubtreeDepth(root->left);
        auto right = checkSubtreeDepth(root->right);
        if (!left.second || !right.second)
            return {-1, false};
        if (abs(left.first - right.first) > 1)
            return {-1, false};
        return {max(left.first, right.first) + 1, true};
    }

    bool isBalanced(TreeNode *root)
    {
        return checkSubtreeDepth(root).second;
    }
};

File: 07_Trees/04_Balanced_Binary_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */

class Solution {

    public int dfs(TreeNode node) {
        if (node == null)
            return 0;
        int left = dfs(node.left);
        int right = dfs(node.right);

        if (left == -1 || right == -1 || Math.abs(left - right) > 1)
            return -1;
        return Math.max(left, right) + 1;
    }

    public boolean isBalanced(TreeNode root) {
        return dfs(root) != -1;
    }
}

class Solution {

    public Pair<Integer, Boolean> dfs(TreeNode node) {
        if (node == null)
            return new Pair<Integer, Boolean>(0, true);
        Pair<Integer, Boolean> left = dfs(node.left);
        Pair<Integer, Boolean> right = dfs(node.right);

        if (!left.getValue() || !right.getValue())
            return new Pair<>(-1, false);
        if (Math.abs(left.getKey() - right.getKey()) > 1)
            return new Pair<>(-1, false);
        return new Pair<>(Math.max(left.getKey(), right.getKey()) + 1, true);
    }

    public boolean isBalanced(TreeNode root) {
        return dfs(root).getValue();
    }
}